torch/distributed/fsdp/_symbolic_trace.py

```diff
@@ -0,0 +1,220 @@
+
+        param_exec_order: a list of parameters ordered based on their execution order.
+
+        _traced_param_set: a set containing all parameters that have been traced.
+
+        module_execution_info_dict: a dict that maps each module to a list of
+            tuple containing a module and a list of named parameters.
+            For a given module, each tuple:
+            1. either contains this module and part of its ``named_parameters`` that will be executed together,
+            2. or contains one of its child modules and all of the child module's ``named_parameters``.
+            The list of tuple is ordered based on the parameter execution order.
+    """
+
+    current_module: torch.nn.Module
+    module_forward_order: List[torch.nn.Module]
+    module_execution_info_dict: Dict[
+        torch.nn.Module,
+        List[Tuple[torch.nn.Module, List[Tuple[str, torch.nn.Parameter]]]],
+    ]
+    param_exec_order: List[torch.nn.Parameter] = field(default_factory=list)
+    _traced_param_set: Set[torch.nn.Parameter] = field(default_factory=set)
+
+
+def _init_execution_info(root_module: torch.nn.Module) -> _ExecutionInfo:
+    """
+    Create an instace of _ExecutionInfo with initialization based on ``root_module``.
+
+    Args:
+        root_module (torch.nn.Module): the module to get the execution information
+            via ``tracer.trace()`` inside ``_patch_tracer``.
+    """
+    return _ExecutionInfo(
+        current_module=root_module,
+        module_forward_order=[root_module],
+        module_execution_info_dict={root_module: []},
+    )
+
+
+def _patched_create_proxy(
+    create_proxy: Callable,
+    execution_info: _ExecutionInfo,
+    params_dict: Dict[str, torch.nn.Parameter],
+    kind: str,
+    target: torch.fx.node.Target,
+    args: Tuple[Any, ...],
+    kwargs: Dict[str, Any],
+    name: Optional[str] = None,
+    type_expr: Optional[Any] = None,
+):
+    """
+    Override of :meth:`~torch.fx.Tracer.create_proxy`.
+    ``Tracer.create_proxy`` is called in symbolic tracing for each leaf function/method/module.
+    This override intercepts the recording of each of these operations and
+    update ``execution_info.module_execution_info_dict``.
+
+    Args:
+        create_proxy (Callable):
+            The ``create_proxy`` function to be patched.
+        execution_info (_ExecutionInfo):
+            Used to repord the execution information.
+        params_dict (Dict[str, torch.nn.Parameter]):
+            A dict that maps each parameter name to the parameter
+        kind, target, args, kwargs, name, type_expr: inputs to the ``create_proxy`` function.
+    """
+    proxy = create_proxy(kind, target, args, kwargs, name, type_expr)
+
+    if kind in ["call_function", "call_method"]:
+        if args is not None:
+            named_params: List[Tuple[str, torch.nn.Parameter]] = []
+            for arg in args:
+                if isinstance(arg, torch.fx.Proxy) and arg.node.target in params_dict:
```

Comment:
Conceptual: When is `arg.node.target` not in `params_dict`?

Reply:
It can also be other types, such as constants used in the function (e.g. axis of a transpose)